sklearn/linear_model/sag.py

```diff
@@ -48,21 +51,28 @@ def get_auto_step_size(max_squared_sum, alpha_scaled, loss, fit_intercept):
     https://hal.inria.fr/hal-00860051/document
     """
     if loss in ('log', 'multinomial'):
-        # inverse Lipschitz constant for log loss
-        return 4.0 / (max_squared_sum + int(fit_intercept)
-                      + 4.0 * alpha_scaled)
+        L = (0.25 * max_squared_sum + int(fit_intercept) + alpha_scaled)
     elif loss == 'squared':
         # inverse Lipschitz constant for squared loss
-        return 1.0 / (max_squared_sum + int(fit_intercept) + alpha_scaled)
+        L = max_squared_sum + int(fit_intercept) + alpha_scaled
     else:
         raise ValueError("Unknown loss function for SAG solver, got %s "
                          "instead of 'log' or 'squared'" % loss)
+    if is_saga:
+        mun = min(2 * n_samples * alpha_scaled, L)
```

Comment:
SAGA original paper: proofs of convergence requires `step_size < 1 / 3L` or `step_size < 1 / (2(L + mu n)`, where `mu` is the strong convexity modulus of the objective. We could use `1 / 3L` but this is more optimal in the low dimensional regime. I will add a reference. By the way SAG use 1 / L whereas the only step size for which proofs are available is `1 / 16 L`. I think this is a sound heuristic but we should add a reference as well. `1 / L` is also a good heuristic for SAGA in most cases, but it actually make one test fail in the test suite :P We could allow the user to specify the step size, but it would complexify the API.